abci/client/committing_client.go

```diff
@@ -22,17 +24,52 @@ type committingClient struct {
 	// Commit
 	// ApplySnapshotChunk
 	mtx *tmsync.RWMutex
+
+	// If called with false, we will wait until the app is ready to serve queries
+	// and deliveries.
+	// CONTRACT: The mtx read lock must be held.
+	//
+	// If called with true, we will indicate that the app is ready.
+	// CONTRACT: The mtx write lock must be held.
+	guardUntilStateIsCommitted func(bool)
+
 	types.Application
 	Callback
 }
 
-func NewCommittingClient(mtx *tmsync.RWMutex, app types.Application) Client {
+// NewCommittingClientGuard uses a condition variable to implement a guard.
+func NewCommittingClientGuard(mtx *tmsync.RWMutex) func(bool) {
+	waitWhileGuardIsActive := sync.NewCond(mtx.RLocker())
+	guardIsActive := true
+	return func(deactivateGuard bool) {
+		if !guardIsActive {
+			// No need for action.  The guard is no longer active.
+			return
+		}
+
+		if deactivateGuard {
+			// Advertise our no-longer-active status.
+			guardIsActive = false
+			waitWhileGuardIsActive.Broadcast()
+			return
+		}
+
+		// We need to wait on the condition.
+		waitWhileGuardIsActive.Wait()
```

Comment:
Add a comment:

```
// It's not necessary to loop testing guardIsActive, as described in the sync.Cond.Wait() API,
// since the variable goes monotonically from true to false.
```

Reply:
Done.